src/join-managed.sol

```diff
@@ -0,0 +1,92 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.6.12;
+
+interface VatLike {
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function dai(address) external view returns (uint256);
+    function gem(bytes32, address) external view returns (uint256);
+    function slip(bytes32, address, int256) external;
+}
+
+interface GemLike {
+    function decimals() external view returns (uint256);
+    function transfer(address, uint256) external returns (bool);
+    function transferFrom(address, address, uint256) external returns (bool);
+}
+
+// For a token that needs join/exit to be managed (like in permissioned vaults)
+
+contract ManagedGemJoin {
+    VatLike public immutable vat;
+    bytes32 public immutable ilk;
+    GemLike public immutable gem;
+    uint256 public immutable dec;
+    uint256 public live;  // Access Flag
+
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+
+    function rely(address usr) external auth { wards[usr] = 1; emit Rely(usr); }
+    function deny(address usr) external auth { wards[usr] = 0; emit Deny(usr); }
+    modifier auth { require(wards[msg.sender] == 1, "ManagedGemJoin/non-authed"); _; }
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event Join(address indexed urn, uint256 amt);
+    event Exit(address indexed urn, address indexed usr, uint256 amt);
+
+    constructor(address vat_, bytes32 ilk_, address gem_) public {
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        gem = GemLike(gem_);
+        dec = GemLike(gem_).decimals();
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    function cage() external auth {
+        live = 0;
+    }
+
+    function mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "ManagedGemJoin/overflow");
+    }
+
+    function join(address urn, uint256 amt) public auth {
+        require(live == 1, "ManagedGemJoin/not-live");
+        uint256 wad = mul(amt, 10 ** (18 - dec));
+        require(int256(wad) >= 0, "ManagedGemJoin/overflow");
+        vat.slip(ilk, urn, int256(wad));
+        require(gem.transferFrom(msg.sender, address(this), amt), "ManagedGemJoin/failed-transfer");
+        emit Join(urn, amt);
+    }
+
+    function exit(address urn, address usr, uint256 amt) public auth {
+        uint256 wad = mul(amt, 10 ** (18 - dec));
+        require(int256(wad) >= 0, "ManagedGemJoin/overflow");
```

Comment:
```suggestion
        require(wad <= 2 ** 255, "ManagedGemJoin/overflow");
```
The other option that makes the maximum amount of `wad` that can actually be `exit`ed.
This is actually the check of the original `GemJoin`.
I'm fine either way, but whatever we use we should try to keep as the standard from now on.

If we accept this change, then I'd also change in join to `wad <= 2 ** 255 - 1` to make them more comparable.

Reply:
Discussed also with @kmbarry1.
Will go with the explicit value check to avoid future solidity compiler casting changes, and to allow the tighter value.